server/src/main/scala/org/http4s/server/armeria/ArmeriaServerBuilder.scala

```diff
@@ -0,0 +1,334 @@
+  override type Self = ArmeriaServerBuilder[F]
+
+  private[this] val logger: Logger = getLogger
+
+  type DecoratorFunction = (HttpService, ServiceRequestContext, HttpRequest) => HttpResponse
+
+  override def bindSocketAddress(socketAddress: InetSocketAddress): Self =
+    copy(socketAddress = socketAddress)
+
+  override def withServiceErrorHandler(serviceErrorHandler: ServiceErrorHandler[F]): Self =
+    copy(serviceErrorHandler = serviceErrorHandler)
+
+  override def resource: Resource[F, ArmeriaServer[F]] =
+    Resource(F.delay {
+      val armeriaServer0 = armeriaServerBuilder
+        .http(socketAddress)
+        .build()
+
+      armeriaServer0.addListener(new ServerListenerAdapter {
+        override def serverStarting(server: BackendServer): Unit = {
+          banner.foreach(logger.info(_))
+
+          val armeriaVersion = Version.get("armeria").artifactVersion()
+
+          logger.info(s"http4s v${BuildInfo.version} on Armeria v${armeriaVersion} started")
+        }
+      })
+      armeriaServer0.start().join()
+
+      val armeriaServer: ArmeriaServer[F] = new ArmeriaServer[F] {
+        lazy val address: InetSocketAddress = {
+          val host = socketAddress.getHostString
+          val port = server.activeLocalPort()
+          new InetSocketAddress(host, port)
+        }
+
+        lazy val server: BackendServer = armeriaServer0
+        lazy val isSecure: Boolean = server.activePort(SessionProtocol.HTTPS) != null
+      }
+
+      armeriaServer -> shutdown(armeriaServer.server)
+    })
+
+  /** Binds the specified `service` at the specified path pattern.
+    * See [[https://armeria.dev/docs/server-basics#path-patterns]] for detailed information of path pattens.
+    */
+  def withHttpService(
+      pathPattern: String,
+      service: (ServiceRequestContext, HttpRequest) => HttpResponse): Self = {
+    armeriaServerBuilder.service(
+      pathPattern,
+      new HttpService {
+        override def serve(ctx: ServiceRequestContext, req: HttpRequest): HttpResponse =
+          service(ctx, req)
+      })
+    this
+  }
+
+  /** Binds the specified [[HttpService]] at the specified path pattern.
+    * See [[https://armeria.dev/docs/server-basics#path-patterns]] for detailed information of path pattens.
+    */
+  def withHttpService(pathPattern: String, service: HttpService): Self = {
+    armeriaServerBuilder.service(pathPattern, service)
+    this
+  }
+
+  /** Binds the specified [[HttpRoutes]] under the specified prefix. */
+  def withHttpRoutes(prefix: String, service: HttpRoutes[F]): Self =
+    withHttpApp(prefix, service.orNotFound)
+
+  /** Binds the specified [[HttpApp]] under the specified prefix. */
+  def withHttpApp(prefix: String, service: HttpApp[F]): Self = {
+    armeriaServerBuilder.serviceUnder(prefix, ArmeriaHttp4sHandler(prefix, service))
+    this
+  }
+
+  /** Decorates all HTTP services with the specified [[DecoratorFunction]]. */
+  def withDecorator(decorator: DecoratorFunction): Self = {
```

Comment:
Do you have a servlet background?  These are roughly equivalent to Filters?  I'm learning Armeria on the fly here.

Reply:
Yes the decorator is similar to Filters but it can transform and lift a `Req` => `Res` type.
https://armeria.dev/docs/server-decorator
Armeria uses the decorator patterns to handle separation of concerns.

JFYI:
- https://armeria.dev/docs/server-decorator
- https://armeria.dev/docs/client-decorator